### PR TITLE
ISSUE-16 - Further unit testing

### PR DESCRIPTION
**Pull request by jmcconnell26**
Adding further unit testing + small refactoring.
Waiting until https://github.com/rust-secure-code/cargo-geiger/pull/125 is merged, and can then merge into this branch.

### CODE REVIEW

**Comment by anderejd** [comment]
#125 is merged now.

**Comment by jmcconnell26** [comment]
Hi @anderejd,

I've merged in the branch to master, and this is now ready for review.
I was wondering if you would have any objections to integrating some code coverage metrics with the CI/CD pipeline?

Thanks,
Josh

**Comment by jmcconnell26** [comment]
I've raised a PR here to add code coverage metrics to the pipeline here https://github.com/rust-secure-code/cargo-geiger/pull/127, which can hopefully be merged in any order!

Thanks,
Josh

File: cargo-geiger/Cargo.toml [hunk 1]
```diff
@@ -20,9 +20,11 @@ cargo-platform = "0.1.1"
 colored = "2.0.0"
 console = "0.11.3"
 env_logger = "0.7.1"
+fake = { version = "2.2", features=["derive"] }
```
[the remark this answers is not in the document]
Comment (jmcconnell26) [reply cargo-geiger/Cargo.toml:23]: I was 50/50 on this library, I really didn't like the fact that it didn't support being added as a dev-dependency for something which was only used in testing.
I came down on the side of including it for how much simpler it made creating large structs in unit tests, for example https://github.com/rust-secure-code/cargo-geiger/pull/126/files#diff-373099d9a5cf5c2a276c879a94c76bdbd5dab804f2dca1ad72776f3b6ea4cb7bR195 for the relative cost of adding derive statement.
If I wasn't to use this, I'd probably create a fixture method with rstest to initialise the default struct, and then inject it into each test.

[the remark this answers is not in the document]
Comment (anderejd) [reply cargo-geiger/Cargo.toml:23]: > I was 50/50 on this library, I really didn't like the fact that it didn't support being added as a dev-dependency for something which was only used in testing.

This alone makes it unfit for use, imho. Making users wait for extra bloat to compile while installing is a bad user experience. Let's go for some extra manually written test code and avoid the `fake` crate.

[the remark this answers is not in the document]
Comment (jmcconnell26) [reply cargo-geiger/Cargo.toml]: Yeah, that definitely makes sense, I've removed it in the last commit.

File: cargo-geiger/src/format/table.rs [hunk 2]
```diff
@@ -313,55 +211,38 @@ mod table_tests {
...
-                total_counter_block: CounterBlock::default(),
-                total_unused_counter_block: CounterBlock::default(),
-            };
-
-        assert_eq!(
-            total_package_counts_none_detected_allows_unsafe
-                .get_total_detection_status(),
-            CrateDetectionStatus::NoneDetectedAllowsUnsafe
+            total_detection_status.get_total_detection_status(),
+            expected_crate_detection_status
```
Comment (anderejd) [inline cargo-geiger/src/format/table.rs:245]: Nice and compact! A clear improvement 👍

File: cargo-geiger/src/format/table.rs [hunk 3]
```diff
@@ -22,128 +31,44 @@ pub const UNSAFE_COUNTERS_HEADER: [&str; 6] = [
 ];
 
 pub fn create_table_from_text_tree_lines(
-    geiger_context: &GeigerContext,
     package_set: &PackageSet,
-    print_config: &PrintConfig,
-    rs_files_used: &HashSet<PathBuf>,
+    table_parameters: &TableParameters,
```
Comment (anderejd) [inline cargo-geiger/src/format/table.rs:35]: Much better 👍

File: cargo-geiger/src/format/table.rs [hunk 4]
```diff
@@ -22,128 +31,44 @@ pub const UNSAFE_COUNTERS_HEADER: [&str; 6] = [
...
+                tree_vines,
+            } => handle_text_tree_line_package(
+                &emoji_symbols,
+                &mut handle_package_parameters,
+                package_id,
+                package_set,
+                &mut table_lines,
+                table_parameters,
+                tree_vines,
+            ),
```
Comment (anderejd) [inline cargo-geiger/src/format/table.rs:71]: Much easier to read with the cases factored out as functions 👍

File: cargo-geiger/Cargo.toml [hunk 5]
```diff
@@ -23,6 +23,7 @@ env_logger = "0.7.1"
 geiger = { path = "../geiger", version = "0.4.5" }
 petgraph = "0.5.1"
 pico-args = "0.3.3"
+rand = "0.7.3"
```
Comment (anderejd) [inline cargo-geiger/Cargo.toml:26]: Can this be a dev-dependency instead?

Comment (jmcconnell26) [reply cargo-geiger/Cargo.toml]: Ahh, apologies, I had meant to remove that, I have taken out in the last commit.

Comment (anderejd) [reply cargo-geiger/Cargo.toml]: Great!

**Pull request merged**